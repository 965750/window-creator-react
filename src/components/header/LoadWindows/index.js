import { connect } from 'react-redux'
import { FormattedMessage } from 'react-intl'

import LoadWindow from '../LoadWindow'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styles from './LoadWindows.module.scss'

class LoadWindows extends Component {
  state = {
    showInfoBox: false,
  }

  componentDidMount() {
    window.addEventListener('click', this.hideInfoBox)
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.hideInfoBox)
  }

  toggleInfoBox = () => {
    this.setState((prevState) => ({
      showInfoBox: !prevState.showInfoBox,
    }))
  }

  hideInfoBox = (e) => {
    if (this.state.showInfoBox && !this.node.contains(e.target)) {
      this.setState({
        showInfoBox: false,
      })
    }
  }

  render() {
    let infoBox
    let savedWindows

    if (this.props.user.windows) {
      savedWindows = this.props.user.windows.map((window, index) => (
        <LoadWindow isLast={this.props.user.windows.length === index + 1} key={window.id} window={window} />
      ))
    }


    if (this.state.showInfoBox) {
      infoBox = (
        <div
          className={`${styles.infoBox} px-2 absolute right-0 border border-gullGray rounded bg-wildSand lowercase`}
        >
          {savedWindows}
        </div>
      )
    }

    return (
      <button
        className="z-40 uppercase mr-5 cursor-pointer"
        onClick={this.toggleInfoBox}
        ref={n => (this.node = n)} // eslint-disable-line
      >
        <span className="hover:text-malachite">
          <FormattedMessage id="Load" defaultMessage="Load" />
        </span>
        {infoBox}
      </button>
    )
  }
}

LoadWindows.propTypes = {
  user: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    isLoaded: PropTypes.bool,
    isEmpty: PropTypes.bool,
    windows: PropTypes.array,
  }),
}

const mapStateToProps = (state) => ({
  user: state.firebase.profile,
})

export default connect(mapStateToProps)(LoadWindows)
