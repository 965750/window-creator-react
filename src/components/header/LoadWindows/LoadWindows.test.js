import '@testing-library/jest-dom/extend-expect'

import { applyMiddleware, compose, createStore } from 'redux';
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { render, cleanup, fireEvent } from '@testing-library/react'

import LoadWindows from './index'
import messages from '../../../messages-i18n'
import thunk from 'redux-thunk';
import React from 'react'

const initialStore = {
  theme: {
    lang: 'pl',
  },
  firebase: {
    profile: {
      firstName: 'mockName',
      lastName: 'mockLastName',
      isLoaded: true,
      isEmpty: false,
      windows: [{
        id: 'mockId',
        save: {
          name: 'mockWindowName'
        }
      }]
    }
  }
}
const reducer = (state = initialStore, action) => {
  switch (action.type) {
    default:
      return state
  }
}

const store = createStore(
  reducer,
  compose(
    applyMiddleware(thunk),
  ),
)

const renderWithRedux = (component) => ({
  ...render(<Provider store={store}>{component}</Provider>),
})

describe('LoadWindows', () => {
  let wrapper
  const initialProps = {}

  const setUp = (props = initialProps) => {
    wrapper = renderWithRedux(
      <IntlProvider
        locale="en"
        messages={messages[store.getState().theme.lang]}
      >
        <LoadWindows {...props} />
      </IntlProvider>,
    )
  }

  afterEach(cleanup)

  it('should match snapshot', () => {
    setUp()

    expect(wrapper.asFragment()).toMatchSnapshot()
  })

  it('should have load option in opened box', () => {
    setUp({
      ...initialStore,
      firebase: {
        profile: {
          firstName: 'mockName',
          lastName: 'mockLastName',
          isLoaded: true,
          isEmpty: false,
          windows: [{
            id: 'mockId',
            save: {
              name: 'mockWindowName'
            }
          }]
        }
      }
    })

    const load = wrapper.getByText(/Wczytaj/)
    fireEvent.click(load)

    expect(wrapper.getByText(/mockWindowName/)).toBeInTheDocument()
  })

  it('should be able to open box then close it', () => {
    setUp({
      ...initialStore,
      firebase: {
        profile: {
          firstName: 'mockName',
          lastName: 'mockLastName',
          isLoaded: true,
          isEmpty: false,
          windows: [{
            id: 'mockId',
            save: {
              name: 'mockWindowName'
            }
          }]
        }
      }
    })

    const load = wrapper.getByText(/Wczytaj/)

    fireEvent.click(load)
    expect(wrapper.queryByTestId('wrapper')).not.toBeNull()

    fireEvent.click(load)
    expect(wrapper.queryByTestId('wrapper')).toBeNull()
  })
})
