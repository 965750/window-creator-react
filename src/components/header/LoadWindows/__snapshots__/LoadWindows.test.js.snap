// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadWindows should match snapshot 1`] = `
<DocumentFragment>
  <button
    class="z-40 uppercase mr-5 cursor-pointer"
  >
    <span
      class="hover:text-malachite"
    >
      Wczytaj
    </span>
  </button>
</DocumentFragment>
`;
